### ISSUE TITLE
Enhancement: allow a whitelist of files to compile

### ISSUE THREAD

**Issue by reharik** [typescript-eslint/typescript-eslint#7012]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

typescript-eslint and the parser look at your project reference and grab the tsconfigs.  They then parse all the files indicated by those tsconfigs. Generally this is the whole project. In a mono-repo this could be several apps. This is good and fine.  But sometimes you only want to lint one file or a few files.  Unfortunately, we must still compile the entire suite of apps before linting that small subset.  

Specific ( maybe only ) example.  In a large legacy app, where you are trying to introduce some semblance of order via eslint, you can not lint the whole repo.  Just as an example our relatively young legacy app :) has like 30k errors.  In fact, often you can not lint an entire file because, as legacy apps tend to do, a file may have 800 lines in it. All of which have lint errors.
 
I have a library ( that really is just a modernization of an older library ) lint-diff-line.  It does a diff against a range, and gets the files that have changed.  Then it gets the line numbers that have changed.  Then it lints the files and filters the messages by the line numbers.
 
In my testing within my legacy app, I would add an error to one file and run it.  It takes 13 seconds to lint that one file. That is because typescript-eslint is compiling our entire application.

The only way around this would be to do some in process manipulating of tsconfig files and even then I'm not sure it would work.  One thing I do have access to during the process is the eslintrc. If I could specify the files that I want to typescript-eslint to compile via a property in the parserOptions that I could access via the eslint api ( or even if I had to parse and modify the eslintrc ) then I could reduces the time to lint a few files by probably 10 seconds.

### Fail

```typescript
compile every file in a mono repo
```

### Pass

```typescript
Only compile a subset of files in a mono repo.
```

### Additional Info

thank you

**Comment by JoshuaKGoldberg** [comment]
Have you tried using [ESLint-specific TSConfigs such as `tsconfig.eslint.json`](https://typescript-eslint.io/architecture/parser#project)? That would allow you to construct TSConfigs that include only the files you care about.

Alternately, you could use the [`program` option](https://typescript-eslint.io/architecture/parser#program) to dynamically create your own program in the ESLint config file.

> large legacy app ... 30k errors

I don't envy you that challenge 😄 good luck!

**Comment by reharik** [comment]
Hi, soooo, it turns out ( as I'm sure y'all already have surmised ), I can't just compile one file. that file has a reference to 10 other files and the ts compiler throws because it can't find those files.  Unless the way typescript-eslint is working is not exactly the same as tsc.  
Unless there's a way to compile files without need for all the files they reference, I think I"m out of luck.  Once you or someone confirms that this is the case I can close the request.
Thanks

**Comment by JoshuaKGoldberg** [comment]
Yup, that's the situation we're in. Ah well!
